### PR TITLE
Let LSP client know which diagnostics are build-only.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This is needed by the LSP client so it can know that if a error is reported by the build and is build-only, that it should stay around till the next build.  Conversely, if an error is reported by the build and is *not* build-only, then it should go away with the next build *or* the next time live diagnostics are retrieved for that document (as the presumption is that the live compiler is responsible for accurate reporting of that info live).

### CODE REVIEW

File: src/EditorFeatures/Core/LanguageServer/AlwaysActivateInProcLanguageClient.cs
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/LanguageServer/AlwaysActivateInProcLanguageClient.cs:107]: by accessing metadata only, we can pull in the info from C#/VB without unnecessary dll loads.

Comment (dibarbet) [reply src/EditorFeatures/Core/LanguageServer/AlwaysActivateInProcLanguageClient.cs:107]: I *think* this should work, but we should run an RPS val build to be certain - build here - https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=7169548&view=results

**Review by kayle — APPROVED** [review]
Looks fine to me. Not sure when we'll have this part done on the LSP client side, but this looks like what I expected.

**Review by dibarbet — APPROVED** [review]

File: src/EditorFeatures/CSharp/LanguageServer/CSharpLspBuildOnlyDiagnostics.cs
Comment (tmat) [inline src/EditorFeatures/CSharp/LanguageServer/CSharpLspBuildOnlyDiagnostics.cs:14]: >LspBuildOnlyDiagnostics [](http://example.com/codeflow?start=5&length=23)

Since these are just strings do we need to MEF export these, instead of just listing all of them (C#+VB) in a single list?

Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharp/LanguageServer/CSharpLspBuildOnlyDiagnostics.cs:14]: Where would that single list live?  I would not want it in a non-C#/VB location as that would not be scalable to other languages.  it would also be leaking specific compiler lang details into a core location.  THis approach keeps C# in C# and VB in VB (And F# in F# if they ever move onto us, for example).  :)

**Comment by CyrusNajmabadi** [comment]
> I think this should work, but we should run an RPS val build to be certain - build here -

Sure.  I have no issue with that :)

File: src/EditorFeatures/CSharp/LanguageServer/CSharpLspBuildOnlyDiagnostics.cs
Comment (tmat) [reply src/EditorFeatures/CSharp/LanguageServer/CSharpLspBuildOnlyDiagnostics.cs:14]: OK, if the intention is to make this available to F# then this is a good approach.

Comment (kayle) [reply src/EditorFeatures/CSharp/LanguageServer/CSharpLspBuildOnlyDiagnostics.cs:14]: If you wanted it in a single list, it'd be shared at the same level as your LSP server. Since VB & C# use the same server instance/config, that code would store the list. If F# shared the same server, it would add it's items to the VB/C# list. If F# used a separate LSP server, it'd manage its own list.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler can i get quick eyes on some comment-only changes to the compiler layer?  Thanks!

**Comment by mavasani** [comment]
This accounts for build-only compiler diagnostics. We would also need to account for build-only analyzer diagnostics (basically diagnostics reported from compilation end actions): https://sourceroslyn.io/#Microsoft.VisualStudio.LanguageServices/TaskList/ExternalErrorDiagnosticUpdateSource.cs,839.

File: src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs:576]: Another approach here would be to move this helper method to a separate file in a separate helper class and then link this file into both the projects? I'll let the compiler team decide their preference.

**Comment by CyrusNajmabadi** [comment]
> We would also need to account for build-only analyzer diagnostics (basically diagnostics reported from compilation end actions

I'm not sure how we could know that data statically.  Is this possible?

**Comment by mavasani** [comment]
> > We would also need to account for build-only analyzer diagnostics (basically diagnostics reported from compilation end actions
> 
> I'm not sure how we could know that data statically. Is this possible?

We could create a constant `ErrorCode[]` in a separate file and link it into both files? It does make it a tad messy, so maybe your approach is cleaner, especially given we have tests validation the lists stay in sync.

**Comment by CyrusNajmabadi** [comment]
@mavasani Sorry, i specifically was asking about the CompilationEndAnalyzers.  Not sure how we could know what diagnostic ids are produced by those statically.  :(  Is that possible?

**Comment by mavasani** [comment]
> @mavasani Sorry, i specifically was asking about the CompilationEndAnalyzers. Not sure how we could know what diagnostic ids are produced by those statically. :( Is that possible?

No, that is not possible. Even the approach we currently take in the IDE is expect the analyzers to add the below special custom tag to the compilation end descriptor, and have a meta-analyzer flagging analyzers that report a compilation-end diagnostic with a descriptor that lacks this tag: https://github.com/dotnet/roslyn/blob/578c44792adc6a37754730fa40c358892aa88839/src/Compilers/Core/Portable/Diagnostic/WellKnownDiagnosticTags.cs#L52-L56

At the minimum, we need loading the analyzers and fetching the supported descriptors and looking at their custom tags to identify if the diagnostic ID is a build-only compilation end diagnostic.

**Comment by CyrusNajmabadi** [comment]
@mavasani 

> No, that is not possible. 

i think we'll have to hold off then on supporting that in LSP.  @kayle can you file a tracking issue about allowing these diagnostic Ids to be provided dynamically, not statically?

**Comment by kayle** [comment]
Yes, added a note to https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_workitems/edit/1612938.

**Comment by dibarbet** [comment]
> @mavasani
> 
> > No, that is not possible.
> 
> i think we'll have to hold off then on supporting that in LSP. @kayle can you file a tracking issue about allowing these diagnostic Ids to be provided dynamically, not statically?

@kayle @CyrusNajmabadi I think we could do this without protocol changes if the client supported dynamic registration of the capabilities perhaps?  We'd raise the [registerCapability ](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#client_registerCapability)event when we load an analyzer with a compilation end tag

File: src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs
Comment (genlu) [inline src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs:576]:
```suggestion
                    // Update src\EditorFeatures\CSharp\LanguageServer\CSharpLspBuildOnlyDiagnostics.cs
```

File: src/EditorFeatures/Core/LanguageServer/AlwaysActivateInProcLanguageClient.cs
Comment (genlu) [inline src/EditorFeatures/Core/LanguageServer/AlwaysActivateInProcLanguageClient.cs:26]:
```suggestion
```

**Review by genlu — APPROVED** [review]
:shipit:

File: src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs:576]: I would prefer this approach.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs:576]: This approach doesn't work for us because it's all based on enums taht themselves are not shared with the IDE.  This space is effectively opaque to us.

**Comment by CyrusNajmabadi** [comment]
> We could create a constant ErrorCode[] in a separate file and link it into both files? It does make it a tad messy, so maybe your approach is cleaner, especially given we have tests validation the lists stay in sync.

This also won't work given that ErrorCode is *itself* internal, and my guess is that the compiler likely wouldn't want all that info to now be copied and linked into the IDE.  But i'm ok with being proven wrong on this.  Ideally the compiler would just expose these build-only diagnostics as a public compilation API.  @333fred do you think you'd be amenable to that?

**Comment by 333fred** [comment]
> Ideally the compiler would just expose these build-only diagnostics as a public compilation API.

Presumably this would be a `string[]` array, since `DiagnosticDescriptor.Id` is `string` today? I'd be ok with that.

**Comment by CyrusNajmabadi** [comment]
>Presumably this would be a string[] array, since DiagnosticDescriptor.Id is string today? I'd be ok with that..

Ok.  Let me look into that.

**Comment by CyrusNajmabadi** [comment]
Fred and I talked offline.  Def issues both with making a public API here, or even with the string[] linking approach.  He's going to report back his thoughts later if we can find a way to navigate this some other way.

**Comment by 333fred** [comment]
After some enumeration of ideas (custom attributes, source generators, updating `generate-compiler-code.ps1`), it feels like the copy and have a test approach is the simplest.

**Comment by CyrusNajmabadi** [comment]
@333fred @dotnet/roslyn-compiler  to review some comment-only additions to the compiler code.  Thanks!

**Comment by CyrusNajmabadi** [comment]
@333fred @cston https://github.com/orgs/dotnet/teams/roslyn-compiler to review some comment-only additions to the compiler code. Thanks!

File: src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs:576]: This needs to be addressed.

**Review by 333fred — COMMENTED** [review]
Looks like Gen has a few unaddressed suggested changes.

File: src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs:576]: Oops.  Missed this.  Sorry Gen!

**Review by 333fred — APPROVED** [review]

**Review by jcouv — APPROVED** [review]
Compiler changes LGTM (iteration 15)

**Pull request merged**
